bpf/generictracer/protocol_kafka.h

```diff
@@ -0,0 +1,443 @@
+// | message_size | correlation_id |
+// +--------------+----------------|
+// |    4B        |       4B.      |
+// +--------------+----------------|
+static __always_inline int kafka_parse_fixup_response_header(const connection_info_t *conn_info,
+                                                             struct kafka_response_hdr *hdr,
+                                                             const unsigned char *data,
+                                                             size_t data_len,
+                                                             u8 direction) {
+    // Try to parse and validate the header first.
+    hdr->message_size = kafka_read_message_size(data, data_len);
+    if (hdr->message_size == -1) {
+        return -1;
+    }
+
+    if (hdr->message_size == (data_len - k_kafka_hdr_message_size)) {
+        // Header is valid and we have the full data, we can proceed.
+        if (kafka_check_response_header_correlation_id(hdr, data + k_kafka_hdr_message_size) < 0) {
+            return -1;
+        }
+        return 0;
+    }
+    // Prepend the header from state data.
+    struct kafka_state_key state_key = {.conn = *conn_info, .direction = direction};
+    struct kafka_state_data *state_data = bpf_map_lookup_elem(&kafka_state, &state_key);
+    if (state_data != NULL && state_data->message_size == data_len) {
+        // Prepend the header from state data.
+        hdr->message_size = state_data->message_size;
+        if (kafka_check_response_header_correlation_id(hdr, data) < 0) {
+            return -1;
+        }
+        return 0;
+    }
+
+    bpf_dbg_printk("failed to parse kafka response header");
+    return -1;
+}
+
+// This is an alternative version of kafka_parse_fixup_response_header that fills the buffer
+// without reading header fields.
+// We are interested only in response, that's why the check on the size is done using
+// the minimum response header size.
+static __always_inline int kafka_read_fixup_response_buffer(const connection_info_t *conn_info,
+                                                            unsigned char *buf,
+                                                            u32 *buf_len,
+                                                            const unsigned char *data,
+                                                            u32 data_len,
+                                                            u8 direction) {
+    u8 offset = 0;
+
+    struct kafka_state_key state_key = {.conn = *conn_info, .direction = direction};
+    struct kafka_state_data *state_data = bpf_map_lookup_elem(&kafka_state, &state_key);
+    if (state_data != NULL && (state_data->message_size == data_len)) {
+        state_data->message_size = bpf_htonl(state_data->message_size);
+        bpf_probe_read(buf, k_kafka_hdr_message_size, (const void *)state_data);
+        offset += k_kafka_hdr_message_size;
+        bpf_map_delete_elem(&kafka_state, conn_info);
```

Comment:
Should this be `&state_key`?

```suggestion
        bpf_map_delete_elem(&kafka_state, &state_key);
```